cmd/fleet/calendar_cron.go

```diff
@@ -0,0 +1,454 @@
+		}
+	}
+}
+
+func getPreferredCalendarEventDate(year int, month time.Month, today int) time.Time {
+	const (
+		// 3rd Tuesday of Month
+		preferredWeekDay = time.Tuesday
+		preferredOrdinal = 3
+	)
+
+	firstDayOfMonth := time.Date(year, month, 1, 0, 0, 0, 0, time.UTC)
+	offset := int(preferredWeekDay - firstDayOfMonth.Weekday())
+	if offset < 0 {
+		offset += 7
+	}
+	preferredDate := firstDayOfMonth.AddDate(0, 0, offset+(7*(preferredOrdinal-1)))
+	if today > preferredDate.Day() {
+		today_ := time.Date(year, month, today, 0, 0, 0, 0, time.UTC)
+		preferredDate = addBusinessDay(today_)
+	}
+	return preferredDate
+}
+
+func addBusinessDay(date time.Time) time.Time {
+	nextBusinessDay := 1
+	switch weekday := date.Weekday(); weekday {
+	case time.Friday:
+		nextBusinessDay += 2
+	case time.Saturday:
+		nextBusinessDay += 1
+	}
+	return date.AddDate(0, 0, nextBusinessDay)
+}
+
+func removeCalendarEventsFromPassingHosts(
+	ctx context.Context,
+	ds fleet.Datastore,
+	calendar fleet.UserCalendar,
+	hosts []fleet.HostPolicyMembershipData,
+) error {
+	for _, host := range hosts {
+		calendarEvent, err := ds.GetCalendarEvent(ctx, host.Email)
+		switch {
+		case err == nil:
+			// OK
+		case fleet.IsNotFound(err):
+			continue
+		default:
+			return fmt.Errorf("get calendar event from DB: %w", err)
+		}
+
+		if err := ds.DeleteCalendarEvent(ctx, calendarEvent.ID); err != nil {
+			return fmt.Errorf("delete db calendar event: %w", err)
+		}
+		if err := calendar.Configure(host.Email); err != nil {
+			return fmt.Errorf("connect to user calendar: %w", err)
+		}
+		if err := calendar.DeleteEvent(calendarEvent); err != nil {
+			return fmt.Errorf("delete calendar event: %w", err)
+		}
+	}
+	return nil
+}
+
+func fireWebhookForHostsWithoutAssociatedEmail(
+	webhookURL string,
+	domain string,
+	hosts []fleet.HostPolicyMembershipData,
+	logger kitlog.Logger,
+) error {
+	// TODO(lucas): We are firing these every 5 minutes...
+	for _, host := range hosts {
+		if err := fleet.FireCalendarWebhook(
+			webhookURL,
+			host.HostID, host.HostHardwareSerial, host.HostDisplayName, nil,
+			fmt.Sprintf("No %s Google account associated with this host.", domain),
+		); err != nil {
+			level.Error(logger).Log(
+				"msg", "fire webhook for hosts without associated email", "err", err,
+			)
+		}
+	}
+	return nil
+}
+
+func generateCalendarEventBody(orgName, hostDisplayName string) string {
+	return fmt.Sprintf(`Please leave your computer on and connected to power.
+
+Expect an automated restart.
+
+%s reserved this time to fix %s.`, orgName, hostDisplayName,
+	)
+}
+
+func isHostOnline(ctx context.Context, ds fleet.Datastore, hostID uint) (bool, error) {
+	hostLite, err := ds.HostLiteByID(ctx, hostID)
+	if err != nil {
+		return false, fmt.Errorf("get host lite: %w", err)
+	}
+	status := (&fleet.Host{
+		DistributedInterval: hostLite.DistributedInterval,
+		ConfigTLSRefresh:    hostLite.ConfigTLSRefresh,
+		SeenTime:            hostLite.SeenTime,
+	}).Status(time.Now())
```

Comment:
Don't we have a common function for this that's already used by the frontend to show online status?

Reply:
Correct. Host details page and hosts list page uses this `Host.Status` method and return it to the UI (the UI then checks the `status` string).